### PR TITLE
Add mentoring notes for Dart Two Fer exercise

### PR DESCRIPTION
**Pull request by Corentin-Leffy**
This PR initializes mentoring notes for the `Two Fer` exercise in Dart to help mentors during their reviews. I started from the C# one and made some adjustments to make it conform to Dart requirements.

### CODE REVIEW

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+The most simple solution is to set the default value for the parameter to `"you"` and use string interpolation:
+
+```dart
+String twofer([String name = "you"]) {
+  return "One for $name, one for me.";
+}
+```
+
+As an alternative, one could use an expression-bodied method:
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]:
```suggestion
Solution using an expression-bodied method:
```

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+```dart
+String twofer([String name = "you"]) => "One for $name, one for me.";
+```
+
+An alternative solution would be to use set the default value to `null` and use the
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]:
```suggestion
Solution which sets the default value to `null` and uses the 
```

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+An alternative solution would be to use set the default value to `null` and use the 
+[null-coalescing operator][reference-null-coalescing-operator]:
+
+```dart
+String twofer([String name = null]) {
+  return "One for ${name ?? "you"}, one for me.";
+}
+```
+
+As an alternative, one could use an expression-bodied method here too:
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]:
```suggestion
Solution with both a default value and an expression-bodied method:
```

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+```
+
+### Common suggestions
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation 
+approach.
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]:
```suggestion
```

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+```dart
+String twofer([String name = "you"]) => "One for $name, one for me.";
+```
+
+An alternative solution would be to use set the default value to `null` and use the 
+[null-coalescing operator][reference-null-coalescing-operator]:
+
+```dart
+String twofer([String name = null]) {
+  return "One for ${name ?? "you"}, one for me.";
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]: I suggest removing this entire solution altogether, as there is a common suggestion to set it to "you".

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+### Common suggestions
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation 
+approach.
+- If they are using `null` as the default value, suggest to use `"you"` instead.
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]: If you give this advice, it should be correct in the solutions up top. Remove the one solution that uses `null` as the default value.

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+### Talking Suggestions
+
+- Many solutions re-assign the name parameter (e.g. `if (name == null) name = "you";`). While there
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]: This reads as a suggestion. Not a talking point.

Instead, "discuss immutability vs mutability" and expand on that.

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation 
+approach.
+- If they are using `null` as the default value, suggest to use `"you"` instead.
+
+### Talking Suggestions
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]:
```suggestion
### Talking points
```

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+
+- Many solutions re-assign the name parameter (e.g. `if (name == null) name = "you";`). While there 
+is technically nothing wrong with this, you could suggest not to re-use the parameter but instead 
+introduce a new variable. This has the main benefit of keeping things immutable: by not overwriting 
+the parameter value, you know its value on every single line in the method: namely its original 
+value. This greatly helps in figuring out the state of a system.
+- Not everyone likes them, but it might be useful to suggest writing the `twofer` method as an
```
Comment (SleeplessByte) [inline tracks/dart/exercises/two-fer/mentoring.md]: This reads as a suggestion, not a talking point.

**Comment by Stargator** [comment]
@SleeplessByte I am not familiar with mentoring notes, the mentors see these, not the students?

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+### Common suggestions
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation.
+- If they are using `null` as the default value, suggest to use `"you"` instead.
+- Not everyone likes them, but it might be useful to suggest writing the `twofer` method as an
```
Comment (Stargator) [inline tracks/dart/exercises/two-fer/mentoring.md:25]: I suggest dropping "Not everyone likes them, but".

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+### Common suggestions
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation.
+- If they are using `null` as the default value, suggest to use `"you"` instead.
+- Not everyone likes them, but it might be useful to suggest writing the `twofer` method as an 
+[expression-bodied method][reference-expression-bodied-method], as it is perfect for these kinds of
```
Comment (Stargator) [inline tracks/dart/exercises/two-fer/mentoring.md:26]: I would at least include a note that these are using the arrow syntax as discussed in the [language tour of Dart](https://dart.dev/guides/language/language-tour#functions).

File: tracks/dart/exercises/two-fer/mentoring.md (hunk 7 above)
Comment (Stargator) [reply tracks/dart/exercises/two-fer/mentoring.md:28]: I good way to rephrase this to to state something like:

> One one best practice is to avoid changing the value of the function parameters.

**Comment by SleeplessByte** [comment]
@Stargator When this file exist, for Dart, on the `two-fer` exercise, mentors will see:

Screencap of the two-fer exercise showing a button to access mentor notes

Pressing it will bring up this content in a modal:

The javascript two-fer mentor notes modal

Which is this page:  https://exercism.io/mentor/exercise_notes?exercise_id=two-fer&track_id=javascript

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation.
+- If they are using `null` as the default value, suggest to use `"you"` instead.
+- It might be useful to suggest writing the `twofer` method as an 
+[expression-bodied method][reference-expression-bodied-method] (also called arrow syntax or `=> expr`),
+ as it is perfect for these kinds of small methods.
+- One one best practice is to avoid changing the value of the function parameters. This has the main
```
Comment (Stargator) [inline tracks/dart/exercises/two-fer/mentoring.md:28]: I made a typo in my suggestion, it should be `One best practice...`

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
+### Reasonable solutions
+
+The most simple solution is to set the default value for the parameter to `"you"` and use string interpolation:
```
Comment (Stargator) [inline tracks/dart/exercises/two-fer/mentoring.md:3]: I would suggest rephrasing it to:

> to set the default value for an optional parameter

File: tracks/dart/exercises/two-fer/mentoring.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+String twofer([String name = "you"]) => "One for $name, one for me.";
+```
+
+### Common suggestions
+
+- If they are using overloaded methods, suggest to use a single method with an 
+[optional parameter][reference-optional-parameter].
+- Instead of using an if-statement, suggest to use a single expression. This will force the student 
+to remove any duplication in the code.
+- Suggest using [string interpolation][reference-string-interpolation] over string concatenation.
```
Comment (Stargator) [inline tracks/dart/exercises/two-fer/mentoring.md:23]: I would add, it is the recommended practice by the Dart development team or that it is a common practice within the Dart community.

**Comment by SleeplessByte** [comment]
Thanks everyone for the effort ❤️
